pkg/utils/deployment_controller.go

```diff
@@ -0,0 +1,114 @@
+	"k8s.io/client-go/kubernetes"
+)
+
+type DeploymentOptions struct {
+	Required       *appsv1.Deployment
+	ControllerName string
+	OpStatus       *operatorapi.OperatorStatus
+	EventRecorder  events.Recorder
+	KubeClient     kubernetes.Interface
+	OperatorClient v1helpers.OperatorClient
+	TargetVersion  string
+	VersionGetter  status.VersionGetter
+	VersionName    string
+}
+
+func CreateDeployment(depOpts DeploymentOptions) (*appsv1.Deployment, error) {
+	lastGeneration := resourcemerge.ExpectedDeploymentGeneration(depOpts.Required, depOpts.OpStatus.Generations)
+	deployment, _, err := resourceapply.ApplyDeployment(depOpts.KubeClient.AppsV1(), depOpts.EventRecorder, depOpts.Required, lastGeneration)
+	if err != nil {
+		// This will set Degraded condition
+		return nil, err
+	}
+
+	// Available: at least one replica is running
+	deploymentAvailable := operatorapi.OperatorCondition{
+		Type: depOpts.ControllerName + operatorapi.OperatorStatusTypeAvailable,
+	}
+	if deployment.Status.AvailableReplicas > 0 {
+		deploymentAvailable.Status = operatorapi.ConditionTrue
+	} else {
+		deploymentAvailable.Status = operatorapi.ConditionFalse
+		deploymentAvailable.Reason = "WaitDeployment"
+		deploymentAvailable.Message = "Waiting for a Deployment pod to start"
+	}
+
+	// Not progressing: all replicas are at the latest version && Deployment generation matches
+	deploymentProgressing := operatorapi.OperatorCondition{
+		Type: depOpts.ControllerName + operatorapi.OperatorStatusTypeProgressing,
+	}
+	if deployment.Status.ObservedGeneration != deployment.Generation {
+		deploymentProgressing.Status = operatorapi.ConditionTrue
+		deploymentProgressing.Reason = "NewGeneration"
+		msg := fmt.Sprintf("desired generation %d, current generation %d", deployment.Generation, deployment.Status.ObservedGeneration)
+		deploymentProgressing.Message = msg
+	} else {
+		if deployment.Spec.Replicas != nil {
+			if deployment.Status.UpdatedReplicas == *deployment.Spec.Replicas {
+				deploymentProgressing.Status = operatorapi.ConditionFalse
+				// All replicas were updated, set the version
+				depOpts.VersionGetter.SetVersion(depOpts.VersionName, depOpts.TargetVersion)
+			} else {
+				msg := fmt.Sprintf("%d out of %d pods running", deployment.Status.UpdatedReplicas, *deployment.Spec.Replicas)
+				deploymentProgressing.Status = operatorapi.ConditionTrue
+				deploymentProgressing.Reason = "WaitDeployment"
+				deploymentProgressing.Message = msg
+			}
+		}
+	}
+
+	depOpts.OpStatus.ReadyReplicas = deployment.Status.ReadyReplicas
+
+	updateGenerationFn := func(newStatus *operatorapi.OperatorStatus) error {
+		if deployment != nil {
+			resourcemerge.SetDeploymentGeneration(&newStatus.Generations, deployment)
+		}
+		return nil
+	}
+
+	if _, _, err := v1helpers.UpdateStatus(depOpts.OperatorClient,
+		v1helpers.UpdateConditionFn(deploymentAvailable),
+		v1helpers.UpdateConditionFn(deploymentProgressing),
+		updateGenerationFn,
+	); err != nil {
+		return nil, err
```

Comment:
I know the idea of this function was to avoid code duplication, but this will make CSO degraded if we can't deploy the problem detector operator. Is that what we want?

Reply:
Why wouldn't we want that? vsphere-problem-detector is not an optional operator.

Reply:
I think it's better to get degraded, it's really not an optional operator.